cmd/server/main.go

```diff
@@ -363,12 +400,11 @@ func watchLoop() {
 	}
 }
 
-func kubeOrDie(config *rest.Config) *kubernetes.Clientset {
-	k8client, err := kubernetes.NewForConfig(config)
-	if err != nil {
-		panic(err.Error())
-	}
-	return k8client
+func clientsOrDie(config *rest.Config) (*kubernetes.Clientset, *sc.Clientset) {
```

Comment:
-1 on this. This would be much better as two functions IMO. In the PR I am working on I will call a function where I only need one of these clients.

Reply:
@darahayes This function is meant to be only called once and the clients themselves are stored in global variables: https://github.com/aerogear/ups-config-operator/blob/master/cmd/server/main.go#L26

Reply:
@darahayes i've changed this to not even return the clients in my current code (not pushed yet)

Reply:
@pb82 in that case I would is there a reason it's not being called in `main()`?

I see a comment about needing to wait until the `ups-config-secret` exists. Currently it is called in the `handleAndroid` and `handleIOS` functions which means the `pushClient` is not initialised until a binding occurs and the service notices a new secret being created. There is a minor flaw in this logic.

What happens if you redeploy the UPS service or if the pod crashes and gets restarted? Those bind actions have already been handled which means the pushClient is never initialised upon the second deploy unless a new binding is created.

In the PR I am working on, I need access to the `pushClient` object and I am running into this issue where it's not initialised. 

I have found that although the secret is not created until after the deployment config (in the unifiedpush-apb), the secret is still ready by the time the config-operator comes up. Which means the `pushClient` could be initialised in `main()`. That being said, I still think the right thing to do would be to update the apb to create the secret before creating the deploymentconfig then call this function in `main()` @pb82 what do you think?

Reply:
@darahayes I think you're mixing two things here. The `pushClient` is different from the k8 or serviceCatalog clients. The latter ones are indeed created in main: https://github.com/aerogear/ups-config-operator/blob/master/cmd/server/main.go#L400

For the push client: until now we only needed it for the bind and unbind actions but in your case you rely on it from the beginning. I would also prefer to create it in `main` but that's currently not possible due to a dependency issue: In the provision role we make a REST call to UPS to create a push application for the namespace: https://github.com/aerogearcatalog/unifiedpush-apb/blob/master/roles/provision-unifiedpush-apb/tasks/provision-ups.yml#L118

The information returned from this call (the push application ID) is stored in the secret. So we can't create the secret before the deployment config.

An alternative: change this to a normal log entry + `return nil`: https://github.com/aerogear/ups-config-operator/blob/master/cmd/server/main.go#L377 Then in your loop, continuously try to acquire the push client and only continue if the function doesn't return `nil` (some caching / checking the global variable in that function would also make sense).

Reply:
@pb82 awesome thanks! Happy enough with this then 😄